storage/replica_raftstorage.go

```diff
@@ -473,8 +469,13 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
 
 	rangeID := r.RangeID
 
-	// First, save the HardState.  The HardState must not be changed
-	// because it may record a previous vote cast by this node.
+	// First, save the HardState. The HardState must not be changed
+	// because it may record a previous vote cast by this node. This is
+	// usually unnecessary because a snapshot is nearly always
+	// accompanied by a new HardState which incorporates both our former
+	// state and new information from the leader, but in the event that
+	// the HardState has not changed, we want to use our own previous
+	// HardState and not one that was transmitted via the snapshot.
 	hardStateKey := keys.RaftHardStateKey(rangeID)
 	hardState, _, err := engine.MVCCGet(r.store.Engine(), hardStateKey, roachpb.ZeroTimestamp, true /* consistent */, nil)
```

Comment:
Just following the old code (nothing else should be changing concurrently here). I'll change it to read from the batch, but note that reading from a batch does not imply reading from a snapshot, so there is no real difference in semantics (since this is the first thing that happens in the new batch).